multi-connector/src/main/scala/it/agilelab/darwin/connector/multi/MultiConnector.scala

```diff
@@ -0,0 +1,324 @@
+  ): (Schema, Array[Byte]) = {
+    val c = connectorInstance(whichEncoding(avroSingleObjectEncoded))
+    c.retrieveSchemaAndAvroPayload(avroSingleObjectEncoded, endianness, getSchema)
+  }
+
+  /**
+    * Extracts the Schema from the ByteBuffer after the method call the ByteBuffer position will be right after the
+    * header.
+    *
+    * @param avroSingleObjectEncoded a ByteBuffer of a Single-Object encoded payload
+    * @return the avro Schema
+    */
+  override def retrieveSchemaAndAvroPayload(
+    avroSingleObjectEncoded: ByteBuffer,
+    endianness: ByteOrder,
+    getSchema: Long => Option[Schema]
+  ): Schema = {
+    val c = connectorInstance(whichEncoding(avroSingleObjectEncoded))
+    c.retrieveSchemaAndAvroPayload(avroSingleObjectEncoded, endianness, getSchema)
+  }
+
+  /**
+    * Extracts the schema ID from the avro single-object encoded at the head of this input stream.
+    * The input stream will have 10 bytes consumed if the first two bytes correspond to the single object encoded
+    * header, or zero bytes consumed if the InputStream supports marking; if it doesn't, the first bytes (up to 2) will
+    * be consumed and returned in the Left part of the Either.
+    *
+    * @param inputStream avro single-object encoded input stream
+    * @param endianness  the endianness that will be used to read fingerprint bytes,
+    *                    it won't affect how avro payload is read, that is up to the darwin user
+    * @return the schema ID extracted from the input data
+    */
+  override def extractId(inputStream: InputStream, endianness: ByteOrder): Either[Array[Byte], Long] = {
+    ConfluentSingleObjectEncoding.extractId(inputStream, endianness).left.flatMap { readBytes =>
+      if (readBytes.isEmpty) {
+        AvroSingleObjectEncodingUtils.extractId(inputStream, endianness)
+      } else {
+        AvroSingleObjectEncodingUtils.extractId(
+          new SequenceInputStream(new ByteArrayInputStream(readBytes), inputStream),
+          endianness
+        )
+      }
+    }
+  }
+
+  /**
+    * Extracts the schema from the avro single-object encoded at the head of this input stream.
+    * The input stream will have 10 bytes consumed if the first two bytes correspond to the single object encoded
+    * header, or zero bytes consumed if the InputStream supports marking; if it doesn't, the first bytes (up to 2) will
+    * be consumed and returned in the Left part of the Either
+    *
+    * @param inputStream avro single-object encoded input stream
+    * @return the schema ID extracted from the input data
+    */
+  override def extractSchema(
+    inputStream: InputStream,
+    endianness: ByteOrder,
+    getSchema: Long => Option[Schema]
+  ): Either[Array[Byte], Schema] = {
+    ConfluentSingleObjectEncoding
+      .extractId(inputStream, endianness)
+      .rightMap(id => id -> (ConfluentSingleObjectEncoded: SingleObjectEncoded))
+      .left
+      .flatMap { readBytes =>
+        if (readBytes.isEmpty) {
```

Comment:
I don't like this implementation but I can see why you did it this way (I make some assumptions here, so correct me if I am wrong):
1. we need to check if the `InputStream` is confluent/single-object-encoded type, but the `extractId()` method checks using a different number of bytes, so we cannot extract them before.
2. we don't know the actual `InputStream` implementation, so depending on the `mark` supported we could end with the `InputStream` without the leading bytes or the original one (which was reset to the start)
3. depending on the `readBytes` value we know if the `InputStream` is missing the initial bytes (`mark` is not supported) and in that case, we prepend them to a new `InputStream`  to have the complete original data available again
3. we cannot move this logic inside the `extractId()` method because we could end up with an `InputStream` implementation different from the one passed as input parameter (we would always return a `SequenceInputStream` if `mark` is not supported).

I don't see a cool way to sort this out, but I think that leaving the responsibility of handling the leading bytes and the `InputStream` to the caller is dangerous. I think that maybe we can play with point 3. above: having another method that handles this internally by returning a `SequenceInputStream` in case the input one does not allow mark could be better (also we would not have code duplications here in this class). If the caller does not want another `InputStream` implementation he/she can still use the current one, but in all other cases (like this one) we can provide a "clearer" way, what do you think?

Reply:
I see you do not like 2 things about this method:
- it's signature (but it's a bit late for this concern, because it's part of the public API of Darwin and I don't think we should change it)
- the low code re-use
I think we should discuss only the latter, do you have any suggestion/proposal related only to this point?

Reply:
If you are talking about the `ConfluentSingleObjectEncoding.extractId()` method I think we can have its signature that way, but we could have another method like `.extractIdAndStream()` that invokes `.extractId()` and performs all the `InputStream` and leading bytes shenanigans to have as output the ID and an `InputStream` that contains the same data as the input one.
In this case we avoid duplicating the `InputStream` handling outside and we can re-use the code. But this is just a small thing, we can leave it for future improvement if you want.

Reply:
I think we can open this as an improvement, but I would not tackle it in this issue